### PR TITLE
Provide PHP 8.1 support

### PR DESCRIPTION
**Pull request by weierophinney**
This patch provides PHP 8.1 support, via the following:

- Adds `~8.1.0` to the PHP requirement constraint
- Adds `#[ReturnTypeWillChange]` attribute to `HydratingIteratorIterator::current` to allow that class to work under PHP 8.1
- Updates tests to run under PHP 8.1; in particular, two tests that were testing for invalid keys needed to change as the behavior of PHP 8.1 when iterating is to raise an error for float values used as keys.

### CODE REVIEW

**Comment by Ocramius** [comment]
I guess we need to run with latest phpunit only...

**Comment by weierophinney** [comment]
> I guess we need to run with latest phpunit only...

I've had success running 8.1 with this PHPUnit series elsewhere - I'm bumping the version to match that on laminas-stdlib, as tests ran there fine.

**Comment by Ocramius** [comment]
Yes, but stdlib does not mock internal PHP classes 😬

**Comment by weierophinney** [comment]
> Yes, but stdlib does not mock internal PHP classes

Good point... so, found that PHPUnit 9.5.5 resolves the issues related to 8.1. Tests are running now; :crossed_fingers: !

File: test/HydratorAwareTraitTest.php [hunk 1]
```diff
@@ -15,6 +15,7 @@ class HydratorAwareTraitTest extends TestCase
 {
     public function testSetHydrator(): void
     {
+        /** @psalm-suppress InvalidScalarArgument False positive */
```
Comment (Ocramius) [inline test/HydratorAwareTraitTest.php:18]: Why are we suppressing this? Isn't `HydratorAwareTrait` a valid trait? :thinking:

File: test/HydratorAwareTraitTest.php [hunk 2]
```diff
@@ -28,6 +29,7 @@ public function testSetHydrator(): void
 
     public function testGetHydrator(): void
     {
+        /** @psalm-suppress InvalidScalarArgument False positive */
```
Comment (Ocramius) [inline test/HydratorAwareTraitTest.php:32]: Why are we suppressing this? Isn't `HydratorAwareTrait` a valid trait? :thinking:

**Review by Ocramius — APPROVED** [review]

File: test/HydratorAwareTraitTest.php (hunk 1 above)
Comment (weierophinney) [reply test/HydratorAwareTraitTest.php:18]: This one was odd.

PHPUnit 9.5 changes the parameter type for the `$traitName` argument to `trait-string`. `HydratorAwareTrait` **is** a trait, and the `::class` usage would evaluate to its fully qualified name, which means it's a trait-string. I think that Psalm just isn't detecting that.

File: test/HydratorAwareTraitTest.php (hunk 2 above)
Comment (weierophinney) [reply test/HydratorAwareTraitTest.php:32]: See above.

File: test/HydratorAwareTraitTest.php (hunk 1 above)
Comment (weierophinney) [reply test/HydratorAwareTraitTest.php:18]: Alternately, the trait-string type just is not understood in the version of Psalm we're using, so it sees `trait-string` and thinks this is a class type or similar, and, either way, cannot resolve it. I'm not sure what version of Psalm that was added.

Comment (Ocramius) [reply test/HydratorAwareTraitTest.php:18]: > trait-string type just is not understood in the version of Psalm we're using

very likely: can we try an upgrade?

Comment (weierophinney) [reply test/HydratorAwareTraitTest.php:18]: So, we're actually at the latest stable release (4.10.0). I'd qualify this as a Psalm bug.

Comment (Ocramius) [reply test/HydratorAwareTraitTest.php:18]: Ok, was worth a shot 👍

**Review by Ocramius — APPROVED** [review]

**Pull request merged**